libnetwork/internal/resolvconf/resolvconf_path.go

```diff
@@ -0,0 +1,57 @@
+package resolvconf
+
+import (
+	"context"
+	"net/netip"
+	"sync"
+
+	"github.com/containerd/log"
+)
+
+const (
+	// defaultPath is the default path to the resolv.conf that contains information to resolve DNS. See Path().
+	defaultPath = "/etc/resolv.conf"
+	// alternatePath is a path different from defaultPath, that may be used to resolve DNS. See Path().
+	alternatePath = "/run/systemd/resolve/resolv.conf"
+)
+
+// For Path to detect systemd (only needed for legacy networking).
+var (
+	detectSystemdResolvConfOnce sync.Once
+	pathAfterSystemdDetection   = defaultPath
+)
+
+// Path returns the path to the resolv.conf file that libnetwork should use.
+//
+// When /etc/resolv.conf contains 127.0.0.53 as the only nameserver, then
+// it is assumed systemd-resolved manages DNS. Because inside the container 127.0.0.53
+// is not a valid DNS server, Path() returns /run/systemd/resolve/resolv.conf
+// which is the resolv.conf that systemd-resolved generates and manages.
+// Otherwise Path() returns /etc/resolv.conf.
+//
+// Errors are silenced as they will inevitably resurface at future open/read calls.
+//
+// More information at https://www.freedesktop.org/software/systemd/man/systemd-resolved.service.html#/etc/resolv.conf
+//
+// TODO(robmry) - alternatePath is only needed for legacy networking ...
+//
+//	Host networking can use the host's resolv.conf as-is, and with an internal
+//	resolver it's also possible to use nameservers on the host's loopback
+//	interface. Once legacy networking is removed, this can always return
+//	defaultPath.
+func Path() string {
+	detectSystemdResolvConfOnce.Do(func() {
+		// Allow tests to point at their own resolv.conf file.
```

Comment:
> resolvconf_test.go has tests which depend on global state set up by other tests. We can't have interdependent unit tests.

I don't think so - nothing's sharing the instance of the daemon that's started by the test, which does this...

```
	d := daemon.New(t, daemon.WithEnvVars("DOCKER_TEST_RESOLV_CONF_PATH="+etchostsPath))
	d.StartWithBusybox(ctx, t, "--experimental", "--ip6tables")
	defer d.Stop(t)
```

Reply:
Did you reply to the right comment? You quoted my review comment in a reply to this unrelated inline comment

Reply:
Oh... right - I didn't notice I'd accidentally left in that comment, so your "how" question looked related to the your comment about global state.

I thought `defaultPath` was the state you were referring to. But, it can't affect other tests, because the daemon's started with the env var set just for this test. Which probably means I missed the point, what global state were you looking at?

(Initially I tried to read the env-var here, to initialise `pathAfterSystemdDetection` to whatever it said. But, I'd forgotten the path is overridden at daemon level via libnetwork options. So, I put the env-var read there instead, and failed to delete that comment. I'll do that now.)

Reply:
> what global state were you looking at?

`golden.NormalizeCRLFToLF`

> (Initially I tried to read the env-var here, to initialise `pathAfterSystemdDetection` to whatever it said. But, I'd forgotten the path is overridden at daemon level via libnetwork options. So, I put the env-var read there instead, and failed to delete that comment. I'll do that now.)

Perfect! I was confused by the orphaned comment.